lib/asn1.ml

```diff
@@ -28,17 +28,36 @@ struct
     }
     [@@deriving ord,eq,show]
 
+    let fix_negative_rsa = function
+      | Ok {n; e} ->
+        if Z.gt n Z.zero then
+          Ok { n; e }
+        else
+          (* A few tools generate keys without leading 0 bit, which gets interpreted
+             as a negative integer. This fix should account for these bad keys. *)
+          let length = n |> Z.numbits |> (fun n -> (n + 3) / 4) in
+          let all_fs = String.make length 'f' |> Z.of_string_base 16 in
+          let positive_n = Z.add all_fs n in
```

Comment:
The `+3` is to round up to the next multiple of 4, since integer division truncates (at least for positive integers). It's equivalent to

```
match n mod 4 with
| 0 -> n / 4
| 1 | 2 | 3 -> (n / 4) + 1
```

Reply:
The `+2` is unexplained as well, it just works around the fact that `numbits` returns `1022` for that 1024-bit RSA key.  That doesn't seem any less sensible than rounding to me.

In any case, the calculation looks obscure to me and so I'd recommend dropping it.  As explained in an earlier comment, I don't think it's worth spending more time on this.